File: test/dijkstra.cpp
#include <Helium/algorithms/dijkstra.h>
#include <Helium/hemol.h>
#include <Helium/smiles.h>
#include <Helium/fileio/moleculefile.h>

#include "test.h"

using namespace Helium;

void test_all_atoms()
{
  HeMol mol;
  try {
    parse_smiles("C1CCCC2C1CCC2", mol);
  } catch(Smiley::Exception &e) {
    std::cerr << e.what();
  }

  Dijkstra<HeMol> d(mol, get_atom(mol, 0));

  COMPARE(0, d.distance(get_atom(mol, 0)));
  COMPARE(1, d.distance(get_atom(mol, 1)));
  COMPARE(2, d.distance(get_atom(mol, 2)));
  COMPARE(3, d.distance(get_atom(mol, 3)));
  COMPARE(2, d.distance(get_atom(mol, 4)));
  COMPARE(1, d.distance(get_atom(mol, 5)));
  COMPARE(2, d.distance(get_atom(mol, 6)));
  COMPARE(3, d.distance(get_atom(mol, 7)));
  COMPARE(3, d.distance(get_atom(mol, 8)));

  std::vector<HeAtom> path;

  path = d.path(get_atom(mol, 0));
  COMPARE(1, path.size());
  COMPARE(0, get_index(mol, path[0]));

  path = d.path(get_atom(mol, 1));
  COMPARE(2, path.size());
  COMPARE(0, get_index(mol, path[0]));
  COMPARE(1, get_index(mol, path[1]));

  path = d.path(get_atom(mol, 2));
  COMPARE(3, path.size());
  COMPARE(0, get_index(mol, path[0]));
  COMPARE(1, get_index(mol, path[1]));
  COMPARE(2, get_index(mol, path[2]));

  path = d.path(get_atom(mol, 3));
  COMPARE(4, path.size());
  COMPARE(0, get_index(mol, path[0]));
  COMPARE(1, get_index(mol, path[1]));
  COMPARE(2, get_index(mol, path[2]));
  COMPARE(3, get_index(mol, path[3]));

  path = d.path(get_atom(mol, 4));
  COMPARE(3, path.size());
  COMPARE(0, get_index(mol, path[0]));
  COMPARE(5, get_index(mol, path[1]));
  COMPARE(4, get_index(mol, path[2]));

  path = d.path(get_atom(mol, 5));
  COMPARE(2, path.size());
  COMPARE(0, get_index(mol, path[0]));
  COMPARE(5, get_index(mol, path[1]));

  path = d.path(get_atom(mol, 6));
  COMPARE(3, path.size());
  COMPARE(0, get_index(mol, path[0]));
  COMPARE(5, get_index(mol, path[1]));
  COMPARE(6, get_index(mol, path[2]));

  path = d.path(get_atom(mol, 7));
  COMPARE(4, path.size());
  COMPARE(0, get_index(mol, path[0]));
  COMPARE(5, get_index(mol, path[1]));
  COMPARE(6, get_index(mol, path[2]));
  COMPARE(7, get_index(mol, path[3]));

  path = d.path(get_atom(mol, 8));
  COMPARE(4, path.size());
  COMPARE(0, get_index(mol, path[0]));
  COMPARE(5, get_index(mol, path[1]));
  COMPARE(4, get_index(mol, path[2]));
  COMPARE(8, get_index(mol, path[3]));
}

struct DijkstraExcludeIndex4Atoms
{
  template<typename MoleculeType, typename AtomType>
  bool operator()(const MoleculeType &mol, AtomType atom) const
  {
    return get_index(mol, atom) != 4;
  }
};


int main()
{
  test_all_atoms();
}

File: test/python/dijkstra.py
import helium
import unittest

SMILES = helium.Smiles()

class TestDijkstra(unittest.TestCase):

    def test_dijkstra(self):
        mol = helium.Molecule()
        SMILES.read('C1CCCC2C1CCC2', mol)

        d = helium.Dijkstra(mol, mol.atom(0))

        self.assertEqual(0, d.distance(mol.atom(0)))
        self.assertEqual(1, d.distance(mol.atom(1)))
        self.assertEqual(2, d.distance(mol.atom(2)))
        self.assertEqual(3, d.distance(mol.atom(3)))
        self.assertEqual(2, d.distance(mol.atom(4)))
        self.assertEqual(1, d.distance(mol.atom(5)))
        self.assertEqual(2, d.distance(mol.atom(6)))
        self.assertEqual(3, d.distance(mol.atom(7)))
        self.assertEqual(3, d.distance(mol.atom(8)))

        path = d.path(mol.atom(0))
        self.assertEqual(1, len(path))
        self.assertEqual(0, path[0].index())

        path = d.path(mol.atom(1))
        self.assertEqual(2, len(path))
        self.assertEqual(0, path[0].index())
        self.assertEqual(1, path[1].index())

        path = d.path(mol.atom(2))
        self.assertEqual(3, len(path))
        self.assertEqual(0, path[0].index())
        self.assertEqual(1, path[1].index())
        self.assertEqual(2, path[2].index())

        path = d.path(mol.atom(3))
        self.assertEqual(4, len(path))
        self.assertEqual(0, path[0].index())
        self.assertEqual(1, path[1].index())
        self.assertEqual(2, path[2].index())
        self.assertEqual(3, path[3].index())

        path = d.path(mol.atom(4))
        self.assertEqual(3, len(path))
        self.assertEqual(0, path[0].index())
        self.assertEqual(5, path[1].index())
        self.assertEqual(4, path[2].index())

        path = d.path(mol.atom(5))
        self.assertEqual(2, len(path))
        self.assertEqual(0, path[0].index())
        self.assertEqual(5, path[1].index())

        path = d.path(mol.atom(6))
        self.assertEqual(3, len(path))
        self.assertEqual(0, path[0].index())
        self.assertEqual(5, path[1].index())
        self.assertEqual(6, path[2].index())

        path = d.path(mol.atom(7))
        self.assertEqual(4, len(path))
        self.assertEqual(0, path[0].index())
        self.assertEqual(5, path[1].index())
        self.assertEqual(6, path[2].index())
        self.assertEqual(7, path[3].index())

        path = d.path(mol.atom(8))
        self.assertEqual(4, len(path))
        self.assertEqual(0, path[0].index())
        self.assertEqual(5, path[1].index())
        self.assertEqual(4, path[2].index())
        self.assertEqual(8, path[3].index())

        self.assertTrue(isinstance(d.infinity(), int))

if __name__ == '__main__':
    unittest.main()
